File: Algorithms/Easy/88.Merge-Sorted-Array/MergeSortedArray.go
package easy

// Time complexity: O(m + n)
// Space complexity: O(m + n)
func merge_1(nums1 []int, m int, nums2 []int, n int) {
	temp := make([]int, len(nums1))
	copy(temp, nums1)

	i, j, k := 0, 0, 0
	for i < m && j < n {
		if temp[i] < nums2[j] {
			nums1[k] = temp[i]
			i++
		} else {
			nums1[k] = nums2[j]
			j++
		}
		k++
	}

	for i < m {
		nums1[k] = temp[i]
		i++
		k++
	}

	for j < n {
		nums1[k] = nums2[j]
		j++
		k++
	}
}

// Time complexity: O(m + n)
// Space complexity: O(1)
func merge_2(nums1 []int, m int, nums2 []int, n int) {
	lastIdx := m + n - 1
	for m > 0 && n > 0 {
		if nums1[m-1] > nums2[n-1] {
			nums1[lastIdx] = nums1[m-1]
			m--
		} else {
			nums1[lastIdx] = nums2[n-1]
			n--
		}
		lastIdx--
	}

	for n > 0 {
		nums1[lastIdx] = nums2[n-1]
		n--
		lastIdx--
	}
}

File: Algorithms/Easy/88.Merge-Sorted-Array/MergeSortedArray.py
# Time complexity: O(m + n)
class Solution:
    # Space complexity: O(m)
    def merge(self, nums1: list[int], m: int, nums2: list[int], n: int) -> None:
        temp = nums1[:]

        i, j = 0, 0
        k = 0
        while i < m and j < n:
            if temp[i] < nums2[j]:
                nums1[k] = temp[i]
                i += 1
            else:
                nums1[k] = nums2[j]
                j += 1
            k += 1

        while i < m:
            nums1[k] = temp[i]
            i += 1
            k += 1

        while j < n:
            nums1[k] = nums2[j]
            j += 1
            k += 1

    # Space complexity: O(1)
    def merge_1(self, nums1: list[int], m: int, nums2: list[int], n: int) -> None:
        last = n + m - 1
        while m > 0 and n > 0:
            if nums1[m - 1] > nums2[n - 1]:
                nums1[last] = nums1[m - 1]
                m -= 1
            else:
                nums1[last] = nums2[n - 1]
                n -= 1
            last -= 1

        while n > 0:
            nums1[last] = nums2[n - 1]
            n -= 1
            last -= 1
